pkg/scheduler/scheduler_test.go

```diff
@@ -1030,3 +1033,145 @@ func TestSchedulerBinding(t *testing.T) {
 		})
 	}
 }
+
+type fakeBinder struct {
```

Comment:
I think the test should be in filtering_test.go and scoring_test.go, this will allow us to avoid all those fakes and will make the tests a lot simpler (can be a case in the already existing tests).

Reply:
Not really. The tests in filtering_test.go and scoring_test.go are based on a statically built snapshot, and then schedule a Pod to verify its result. While this test needs to test how the 3 goroutines (main scheduling, cache, schedulingQ) work together. So here I have to use a fake binder to set the pod.spec.nodeName manually, to mimic the v1.Binding behavior.

Reply:
> The tests in filtering_test.go and scoring_test.go are based on a statically built snapshot, and then schedule a Pod to verify its result.

And that is enough to test whether or not the PodTopologySpread logic excludes terminating pods. Why do you need to create a full scheduler and all three routines to test that?

Reply:
First of all, we should have a unit test for just "terminated pods" in filtering_test.go and scoring_test.go.

Now, this complex scenario that spans through a few cycles is more of integration test. A better place for it would be `test/integration/scheduler/priorities_test.go`. And then we don't need fakes. But... this might be an overkill.

That said, do we have any tests that make sure our cache stays up to date with apiserver updates? That seems useful, but definitely outside of the scope of this PR.

Reply:
It looks like we are lacking a bunch of unit tests for eventhandlers.go :(
But maybe we should handle this in integration directly. If you agree, I'll open an issue.

Reply:
> And that is enough to test whether or not the PodTopologySpread logic excludes terminating pods. Why do you need to create a full scheduler and all three routines to test that?

It's true that from the code's coverage point of view, a unit test in filtering_test.go and scoring_test.go is enough. But that wouldn't reproduce the issue - that'd be merely based our debugging result and lack the whole context. I don't believe that'd be sufficient. And as I mentioned in https://github.com/kubernetes/kubernetes/pull/87845#discussion_r375111677, once we get a reporting saying that a new Pod's scheduling cycle can begin before a previous one's caching cycle, this test can quickly detect that and verify potential fix.
Also, we do have similar UTs exercising the full scheduler, but not fully adhere to the new framework. This test can work as a sample to refactor those tests.

An e2e test can work the same as this UT, but that's expensive.

> First of all, we should have a unit test for just "terminated pods" in filtering_test.go and scoring_test.go.

I can make it later once we got to a consensus on the necessity of whether we need this UT.

> Now, this complex scenario that spans through a few cycles is more of integration test.

I thought about an integration test also. However, it's challenging to simulate deleting a Pod while triggering an update event. As you know, in an integration test, there is no kubelet running, you can only "delete" a Pod using deletionGracePeriod=0.

> It looks like we are lacking a bunch of unit tests for eventhandlers.go :(

Probably, we can give a try. But according to my experiences, it'd easier to be covered in UT, using fake clientset and watch.

Reply:
> I thought about an integration test also. However, it's challenging to simulate deleting a Pod while triggering an update event. As you know, in an integration test, there is no kubelet running, you can only "delete" a Pod using deletionGracePeriod=0.

What if you do two separate `cs.CoreV1().Pods(ns).Delete`? One with a grace period and another without it. In that meantime, you wait for a scheduling cycle.

I still think we should have a true "unit" test that is not concerned about order of events, just the fact that terminated pods should be excluded. The thing is, I don't think we had a bug, but rather a new requirement. And this is independent of order of events, which is what we initially thought. It's just the deployments controller creating a new pod early (which is good for availability). We don't have bugs on how we are updating the pods when a pod is terminating. (We should have unit tests for this though, as I said before). But, if we truly believe that we need to cover this, we should have an E2E test (with the controllers and kubelets).

If we plan to add more "UTs" exercising a full scheduler, maybe we should have a folder just for them with some utilities (like the fake binder) in their own package.

Reply:
> What if you do two separate cs.CoreV1().Pods(ns).Delete? One with a grace period and another without it. In that meantime, you wait for a scheduling cycle.

I guess the first one (with grace period) wouldn't work at all, but we can give a try, in terms of integration test.

> I still think we should have a true "unit" test that is not concerned about order of events, just the fact that terminated pods should be excluded.

It's what you mentioned as a pure UT in filtering_test.go & scoring_test.go, right? Sure, I can add it.

> We don't have bugs on how we are updating the cache when a pod is terminating (We should have unit tests for this though, as I said before)

There _may_ be a bug: as I mentioned earlier, we don't guarantee the cache's handling cycle on an old pod can finish **every time** before the new pod's scheduling cycle, even if controller manager and API server can guarantee the event of old pod will always be dispatched before the new pod. The case is kind like this:

- T1: old pod update event comes in. Then cache goroutine gets an update on old pod's update (spec.deletionTimestamp != nil)
- T2: new pod event comes in. Then schedulingQ pushes a new pod and Pop() it
- T3: main scheduling goroutine starts to do snapshot() on cache, and starts to schedule the new pod
- T4: cache goroutine finishes its cycle on old pod's update

I'm not sure in a real case it can happen, but in a UT, this can simulated b/c in T1, it doesn't immediately lock the cache. So I think the possibility is low, and in this UT, I intentionally wait for the cache cycle to be finished and push the new pod.

Reply:
> It's what you mentioned as a pure UT in filtering_test.go & scoring_test.go, right? Sure, I can add it.

Yes, we should have that regardless, in addition to the one for pod namespace.

> There may be a bug: as I mentioned earlier, we don't guarantee the cache's handling cycle ...

We can certainly add tests to verify that the cache is updated properly and that whatever state in the snapshot is consistent, but all of this isn't related to how we implement the exclusion of terminating pods in the Plugin.

Reply:
> It's what you mentioned as a pure UT in filtering_test.go & scoring_test.go, right? Sure, I can add it.

Yes

> So I think the possibility is low, and in this UT, I intentionally wait for the cache cycle to be finished and push the new pod.

So you are saying this UT is not covering the sequence you described?

In any case, that should be covered in a separate unit test for the cache. Basically, a test that adds events and immediately after produces a snapshot, which should be consistent.

Reply:
> Yes, we should have that regardless, in addition to the one for pod namespace.

Will do that.

> We can certainly add tests to verify that the cache is updated properly and that the snapshot state is consistent

It's just that it's _possible_ the cache cycle started, but hasn't reached to locking and update the cache in place; while the scheduling cycle starts, and lock the (stale) cache and schedule a new pod.

> but all of this isn't related to how we implement the exclusion of terminating pods in the Plugin.

Nope, this is not covered in this UT. Actually, that needs a "fix".

> So you are saying this UT is not covering the sequence you described?

It's not - it's intentionally waiting for cache cycle to be finished:

```go
// Simulate a Pod deletion operation - either by user or controller manager.
fakeWatch.Modify(p1Copy)
// Ensure p1Copy has been handled in cache.
wait.PollImmediate(100*time.Millisecond, time.Second, func() (bool, error) {
	p1InCache, err := cache.GetPod(p1Copy)
	return p1InCache.GetDeletionTimestamp() != nil, err
})
```

> In any case, that should be covered in a separate unit test for the cache. Basically, a test that adds events and immediately after produces a snapshot, which should be consistent.

We can pursue that, in another test along with a "fix".

Reply:
for a separate PR :)

Other than that, for this feature, I think we are good to go with a pure unit test.

Reply:
> for a separate PR :)

SGTM. Will extract the complex UT in this PR to a separate one, and only focus on verifying the cache/snapshot consistency.